File: app/src/main/java/com/hover/stax/transactions/TransactionDetailsFragment.java
package com.hover.stax.transactions;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.text.Html;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.RelativeLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatButton;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.RecyclerView;

import com.hover.sdk.actions.HoverAction;
import com.hover.sdk.transactions.Transaction;
import com.hover.sdk.transactions.TransactionContract;
import com.hover.stax.R;
import com.hover.stax.bounties.BountyActivity;
import com.hover.stax.contacts.StaxContact;
import com.hover.stax.databinding.FragmentTransactionBinding;
import com.hover.stax.navigation.NavigationInterface;
import com.hover.stax.utils.DateUtils;
import com.hover.stax.utils.UIHelper;
import com.hover.stax.utils.Utils;

import org.json.JSONException;
import org.json.JSONObject;

public class TransactionDetailsFragment extends Fragment implements NavigationInterface {

    final public static String TAG = "TransDetailsFragment";
    final public static String SHOW_BOUNTY_SUBMIT = "bounty_submit_button";

    private TransactionDetailsViewModel viewModel;

    private FragmentTransactionBinding binding;

    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        viewModel = new ViewModelProvider(requireActivity()).get(TransactionDetailsViewModel.class);
        JSONObject data = new JSONObject();
        try {
            assert getArguments() != null;
            data.put("uuid", getArguments().getString(TransactionContract.COLUMN_UUID));
        } catch (JSONException e) {
            Utils.logErrorAndReportToFirebase(TAG, e.getMessage(), e);
        }

        Utils.logAnalyticsEvent(getString(R.string.visit_screen, getString(R.string.visit_transaction)), data, requireContext());

        binding = FragmentTransactionBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        startObservers();
        setUssdSessionMessagesRecyclerView();
        viewModel.setTransaction(getArguments().getString(TransactionContract.COLUMN_UUID));
    }

    private void setUssdSessionMessagesRecyclerView() {
        setUSSDMessagesRecyclerView();
        setSmsMessagesRecyclerView();
    }

    private void startObservers() {
        viewModel.getTransaction().observe(getViewLifecycleOwner(), this::showTransaction);
        viewModel.getAction().observe(getViewLifecycleOwner(), this::showActionDetails);
        viewModel.getContact().observe(getViewLifecycleOwner(), this::updateRecipient);
    }

    private void setupRetryBountyButton() {
        RelativeLayout bountyButtonsLayout = binding.retrySubmit.bountyRetryButtonLayoutId;
        AppCompatButton retryButton = binding.retrySubmit.btnRetry;

        bountyButtonsLayout.setVisibility(View.VISIBLE);
        retryButton.setOnClickListener(this::retryBountyClicked);
    }

    private void showTransaction(StaxTransaction transaction) {
        if (transaction != null) {
            if (transaction.isRecorded()) setupRetryBountyButton();
            updateDetails(transaction);
            showNotificationCard(transaction.isRecorded() || transaction.status.equals(Transaction.PENDING));
            if (transaction.isRecorded() && viewModel.getAction().getValue() != null)
                updateNotificationCard(viewModel.getAction().getValue());
        }
    }

    @SuppressLint("SetTextI18n")
    private void updateDetails(StaxTransaction transaction) {
        binding.transactionDetailsCard.setTitle(transaction.description);
        binding.detailsRecipientLabel.setText(transaction.transaction_type.equals(HoverAction.RECEIVE) ? R.string.sender_label : R.string.recipient_label);
        binding.detailsAmount.setText(transaction.getDisplayAmount());
        binding.detailsDate.setText(DateUtils.humanFriendlyDate(transaction.initiated_at));

        if (transaction.confirm_code != null && !transaction.confirm_code.isEmpty())
            binding.detailsTransactionNumber.setText(transaction.confirm_code);
        else
            binding.detailsTransactionNumber.setText(transaction.uuid);

        if (transaction.isRecorded()) hideDetails();
    }

    private void hideDetails() {
        binding.amountRow.setVisibility(View.GONE);
        binding.recipientRow.setVisibility(View.GONE);
        binding.recipAccountRow.setVisibility(View.GONE);
    }

    private void showActionDetails(HoverAction action) {
        if (action != null) {
            binding.detailsNetwork.setText(action.from_institution_name);
            if (viewModel.getTransaction().getValue() != null && viewModel.getTransaction().getValue().isRecorded())
                updateNotificationCard(action);
        }
    }

    private void showNotificationCard(boolean show) {
        binding.notificationCard.setVisibility(show ? View.VISIBLE : View.GONE);
    }

    @SuppressLint("ResourceAsColor")
    private void updateNotificationCard(HoverAction action) {
        binding.notificationCard.setBackgroundColor(action.bounty_is_open ? R.color.pending_brown : R.color.muted_green);
        binding.notificationCard.setTitle(R.string.checking_your_flow);
        binding.notificationCard.setIcon(action.bounty_is_open ? R.drawable.ic_warning : R.drawable.ic_check);
        binding.notificationDetail.setText(Html.fromHtml(action.bounty_is_open ? getString(R.string.bounty_flow_pending_dialog_msg) : getString(R.string.flow_done_desc)));
    }

    private void updateRecipient(StaxContact contact) {
        if (contact != null)
            binding.detailsRecipient.setContact(contact);
    }

    private void setUSSDMessagesRecyclerView() {
        RecyclerView messagesView = binding.convoRecyclerView;
        messagesView.setLayoutManager(UIHelper.setMainLinearManagers(requireActivity()));
        messagesView.setHasFixedSize(true);

        viewModel.getMessages().observe(getViewLifecycleOwner(), ussdCallResponses -> {
            if (ussdCallResponses != null)
                messagesView.setAdapter(new MessagesAdapter(ussdCallResponses));
        });
    }

    private void setSmsMessagesRecyclerView() {
        RecyclerView smsView = binding.smsRecyclerView;
        smsView.setLayoutManager(UIHelper.setMainLinearManagers(requireActivity()));
        smsView.setHasFixedSize(true);
        viewModel.getSms().observe(getViewLifecycleOwner(), smses -> {
            if (smses != null) smsView.setAdapter(new MessagesAdapter(smses));
        });
    }

    private void retryBountyClicked(View v) {
        if (viewModel.getTransaction().getValue() != null)
            ((BountyActivity) requireActivity()).retryCall(viewModel.getTransaction().getValue().action_id);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();

        binding = null;
    }
}

File: app/src/main/java/com/hover/stax/transactions/TransactionDetailsFragment.kt
package com.hover.stax.transactions

import android.annotation.SuppressLint
import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.text.method.LinkMovementMethod
import android.view.LayoutInflater
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.graphics.drawable.RoundedBitmapDrawableFactory
import androidx.core.os.bundleOf
import androidx.core.text.HtmlCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.navigation.fragment.findNavController
import androidx.navigation.fragment.navArgs
import com.hover.sdk.actions.HoverAction
import com.hover.sdk.api.Hover
import com.hover.sdk.transactions.Transaction
import com.hover.stax.ApplicationInstance
import com.hover.stax.R
import com.hover.stax.contacts.StaxContact
import com.hover.stax.databinding.FragmentTransactionBinding
import com.hover.stax.home.MainActivity
import com.hover.stax.utils.AnalyticsUtil.logAnalyticsEvent
import com.hover.stax.utils.AnalyticsUtil.logErrorAndReportToFirebase
import com.hover.stax.utils.DateUtils
import com.hover.stax.utils.NavUtil
import com.hover.stax.utils.UIHelper
import com.hover.stax.utils.UIHelper.loadImage
import com.hover.stax.utils.Utils
import com.squareup.picasso.Picasso
import com.squareup.picasso.Target
import org.json.JSONException
import org.json.JSONObject
import org.koin.androidx.viewmodel.ext.android.viewModel
import timber.log.Timber


class TransactionDetailsFragment : Fragment() {

    private val viewModel: TransactionDetailsViewModel by viewModel()
    private var _binding: FragmentTransactionBinding? = null
    private val binding get() = _binding!!
    private val retryCounter = ApplicationInstance.txnDetailsRetryCounter

    private val args: TransactionDetailsFragmentArgs by navArgs()

    private lateinit var childFragManager: FragmentManager

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        viewModel.setTransaction(args.uuid)

        val data = JSONObject()
        try {
            data.put("uuid", args.uuid)
        } catch (e: JSONException) {
            logErrorAndReportToFirebase(TransactionDetailsFragment::class.java.simpleName, e.message!!, e)
        }

        logAnalyticsEvent(getString(R.string.visit_screen, getString(R.string.visit_transaction)), data, requireContext())
        _binding = FragmentTransactionBinding.inflate(inflater, container, false)

        childFragManager = childFragmentManager

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        startObservers()

        binding.transactionDetailsCard.setOnClickIcon { NavUtil.navigate(findNavController(), TransactionDetailsFragmentDirections.actionTxnDetailsFragmentToNavigationHome()) }
        binding.primaryStatus.viewLogText.setOnClickListener { showUSSDLog() }
        with(binding.infoCard.detailsStaxUuid.content) { setOnClickListener { Utils.copyToClipboard(this.text.toString(), requireContext()) } }
        with(binding.infoCard.detailsServiceId.content) { setOnClickListener { Utils.copyToClipboard(this.text.toString(), requireContext()) } }
    }

    private fun showUSSDLog() {
        val log = USSDLogBottomSheetFragment.newInstance(args.uuid)
        log.show(childFragManager, USSDLogBottomSheetFragment::class.java.simpleName)
    }

    private fun startObservers() = with(viewModel) {
        transaction.observe(viewLifecycleOwner) { showTransaction(it) }
        action.observe(viewLifecycleOwner) { it?.let { showActionDetails(it) } }
        contact.observe(viewLifecycleOwner) { updateRecipient(it) }
    }

    private fun setupContactSupportButton(id: String, contactSupportTextView: TextView) {
        contactSupportTextView.setText(R.string.email_support)
        contactSupportTextView.setOnClickListener {
            resetTryAgainCounter(id)
            val deviceId = Hover.getDeviceId(requireContext())
            val subject = "Stax Transaction failure - support id- {${deviceId}}"
            Utils.openEmail(subject, requireActivity())
        }
    }

    private fun updateRetryCounter(id: String) {
        val currentCount: Int = if (retryCounter[id] != null) retryCounter[id]!! else 0
        retryCounter[id] = currentCount + 1
    }

    private fun resetTryAgainCounter(id: String) {
        retryCounter[id] = 0
    }

    private fun shouldContactSupport(id: String): Boolean = if (retryCounter[id] != null) retryCounter[id]!! >= 3 else false

    private fun showTransaction(transaction: StaxTransaction?) {
        if (transaction != null) {
            if (transaction.isRecorded)
                setupRetryBountyButton()
            else if (transaction.status == Transaction.FAILED) {
                val button = showButtonToClick()
                if (shouldContactSupport(transaction.action_id))
                    setupContactSupportButton(transaction.action_id, button)
                else
                    retryTransactionClicked(transaction, button)
            } else binding.secondaryStatus.transactionRetryButtonLayoutId.visibility = GONE
            updateDetails(transaction)
        }
    }

    private fun showButtonToClick(): TextView {
        val transactionButtonsLayout = binding.secondaryStatus.transactionRetryButtonLayoutId
        val retryButton = binding.secondaryStatus.btnRetryTransaction
        transactionButtonsLayout.visibility = VISIBLE
        return retryButton
    }

    private fun setupRetryBountyButton() {
        val bountyButtonsLayout = binding.secondaryStatus.transactionRetryButtonLayoutId
        val retryButton = binding.secondaryStatus.btnRetryTransaction
        bountyButtonsLayout.visibility = VISIBLE
        retryButton.setOnClickListener { retryBountyClicked() }
    }

    private fun retryBountyClicked() {
        viewModel.transaction.value?.let {
            (requireActivity() as MainActivity).retryCall(it.action_id)
        }
    }

    private fun retryTransactionClicked(transaction: StaxTransaction, retryButton: TextView) {
        retryButton.setOnClickListener {
            updateRetryCounter(transaction.action_id)
            if (transaction.isBalanceType) (requireActivity() as MainActivity).reBuildHoverSession(transaction)
            else (requireActivity() as MainActivity).navigateTransferAutoFill(transaction.transaction_type, transaction.uuid)
        }
    }

    private fun shouldShowNewBalance(transaction: StaxTransaction): Boolean {
        return !transaction.isBalanceType && !transaction.balance.isNullOrEmpty() && transaction.isSuccessful
    }

    @SuppressLint("SetTextI18n")
    private fun updateDetails(transaction: StaxTransaction) {
        val title = when (transaction.transaction_type) {
            HoverAction.P2P -> getString(R.string.send_money)
            HoverAction.AIRTIME -> getString(R.string.buy_airtime)
            else -> transaction.transaction_type
        }
        binding.transactionDetailsCard.setTitle(title)

        setDetailsData(transaction)
        setVisibleDetails(transaction)
        updateDetailsRequiringAction(viewModel.action.value, viewModel.transaction.value)
        updateStatus(viewModel.action.value, transaction)
    }

    private fun setDetailsData(transaction: StaxTransaction) = with(binding.infoCard) {
        if (shouldShowNewBalance(transaction)) {
            binding.primaryStatus.newBalance.apply {
                text = getString(R.string.new_balance, transaction.displayBalance)
                visibility = VISIBLE
            }
        }
        detailsRecipientLabel.setText(if (transaction.transaction_type == HoverAction.RECEIVE) R.string.sender_label else R.string.recipient_label)
        detailsAmount.text = transaction.displayAmount
        detailsDate.text = DateUtils.humanFriendlyDateTime(transaction.updated_at)
        detailsServiceId.content.text = transaction.confirm_code
        detailsStaxUuid.content.text = transaction.uuid
        detailsStaxStatus.apply {
            text = transaction.fullStatus.getPlainTitle(requireContext())
            setCompoundDrawablesWithIntrinsicBounds(0, 0, transaction.fullStatus.getIcon(), 0)
        }

        detailsStaxReason.text = transaction.fullStatus.getReason()
        transaction.fee?.let { binding.infoCard.detailsFee.text = Utils.formatAmount(it) }
    }

    private fun setVisibleDetails(transaction: StaxTransaction) = with(binding.infoCard) {
        reasonRow.visibility = if (transaction.isFailed) VISIBLE else GONE
        amountRow.visibility = if (transaction.isRecorded || transaction.transaction_type == HoverAction.BALANCE) GONE else VISIBLE
        recipientRow.visibility = if (transaction.isRecorded || transaction.transaction_type == HoverAction.BALANCE) GONE else VISIBLE
        recipAccountRow.visibility = if (transaction.isRecorded || transaction.transaction_type == HoverAction.BALANCE) GONE else VISIBLE
        serviceIdRow.visibility = if (transaction.isRecorded || transaction.confirm_code.isNullOrBlank()) GONE else VISIBLE
        feeRow.visibility = if (transaction.fee == null) GONE else VISIBLE
    }

    private fun updateDetailsRequiringAction(action: HoverAction?, transaction: StaxTransaction?) {
        if (action != null && transaction != null) {
            binding.infoCard.detailsStaxType.text = transaction.fullStatus.getDisplayType(requireContext(), action)
            binding.infoCard.detailsStaxAccount.text = action.from_institution_name
            binding.infoCard.detailsFeeLabel.text = getString(R.string.transaction_fee, action.from_institution_name)
        }
    }

    private fun showActionDetails(action: HoverAction) {
        binding.infoCard.detailsNetwork.text = action.from_institution_name
        updateStatus(action, viewModel.transaction.value)
        updateDetailsRequiringAction(viewModel.action.value, viewModel.transaction.value)
    }

    private fun updateStatus(action: HoverAction?, transaction: StaxTransaction?) {
        if (action != null && transaction != null) {
            setPrimaryStatus(transaction)
            setSecondaryStatus(action, transaction)
        }
    }

    private fun setPrimaryStatus(transaction: StaxTransaction?) {
        transaction?.let {
            var textValue = transaction.fullStatus.getTitle(requireContext())
            if (transaction.fullStatus.getReason().isNotEmpty()) textValue = textValue + ": " + transaction.fullStatus.getReason()

            binding.primaryStatus.apply {
                statusText.text = textValue
                statusIcon.setImageResource(transaction.fullStatus.getIcon())
            }
        }
    }

    private fun setSecondaryStatus(action: HoverAction?, transaction: StaxTransaction?) {
        transaction?.let {
            if (transaction.isSuccessful) binding.secondaryStatus.root.visibility = GONE
            else {
                binding.secondaryStatus.root.visibility = VISIBLE
                binding.secondaryStatus.statusText.apply {
                    val content = transaction.fullStatus.getStatusDetail(action, viewModel.messages.value?.last(), viewModel.sms.value, requireContext())
                    text = HtmlCompat.fromHtml(content, HtmlCompat.FROM_HTML_MODE_LEGACY)
                    movementMethod = LinkMovementMethod.getInstance()
                }
                if (transaction.isFailed) action?.let {
                    binding.secondaryStatus.statusIcon.loadImage(this@TransactionDetailsFragment, getString(R.string.root_url).plus(it.from_institution_logo))
                }
                else binding.secondaryStatus.statusIcon.visibility = GONE
            }
        }
    }

    private fun updateRecipient(contact: StaxContact?) {
        if (contact != null) {
            binding.infoCard.detailsRecipient.setContact(contact)
        } else binding.infoCard.detailsRecipient.setTitle(getString(R.string.self_choice))
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}